README.md
node-digikey
![logo](http://www.digikey.com/-/media/Images/Header/logo_dk.png)
====

node-digikey is a node.js wrapper for the DigiKey REST API.

Official [DigiKey REST API documentation](https://api-portal.digikey.com/).

##Install

    npm install node-digikey

##Quickstart

```js
var cli = require('node-digikey').createV1('myapikey');

##Features

* HTTPS by default
* Simple node.js style callbacks(err, result)

##TESTS

To run the tests use

    npm test

##API

TODO


##License

Copyright (C) 2017 Rob Sawyer

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
node-digikey
====

###v0.0.1 (2017/02/24)
* Initial release


tests/apiSpec.js
var assert = require('chai').assert;
var sinon = require('sinon');
var PassThrough = require('stream').PassThrough;
var http = require('http');
var request = require('request');

require('dotenv').config();
require('request-debug')(request);

var api = require('../index.js');

var cli = undefined;

describe('DigiKey Api', function() {
	beforeEach(function() {
		this.request = sinon.stub(http, 'request');
	});

	afterEach(function() {
		http.request.restore();
	});

	//We will place our tests cases here
	it('should create an instance of the cli', function(done) {
		// var expected = {
		// 	"access_token": "weoiaslkjfoiw32/12#kd",
		// 	"refresh_token": "aslidf390-sdl/jliDLSksli",
		// 	"token_type": "Bearer",
		// 	"expires_in": 2592000,
		// 	"expires_at": 2592000
		// };
		// var response = new PassThrough();
		// response.write(JSON.stringify(expected));
		// response.end();
		//
		// var request = new PassThrough();

		// this.request
		// 	.callsArgWith(1, response)
		// 	.returns(request);

		// Get a code
		// Error: https://my-new-app.example.com/code?error=access_denied
		// Success: https://my-new-app.example.com/code?code=6513183215H5465sdlkjKils

		//NOTE: In order to use the tests. You'll need to generate your own authorization token.
		cli = api.createV1(process.env.DIGIKEY_AUTH_TOKEN);
		if(!process.env.DIGIKEY_AUTH_TOKEN){
			var err = 'You must provide a valid authorization token. See https://api-portal.digikey.com/start for more.';
			done(err);
		}
		assert.isObject(cli, 'cli is an object');
		done();
	});

	//We will place our tests cases here
	it('should get search results', function(done) {
		// args, filters, callback
		cli.basicSearch({},{},function(err, results){
			console.log(results);
			if(err){
				done(err);
			}
			assert.isAbove(results, 0, 'result length above 0');
			done();
		});
	});

	// it('should send post params in request body', function() {
	// 	var params = { foo: 'bar' };
	// 	var expected = JSON.stringify(params);
	//
	// 	var request = new PassThrough();
	// 	var write = sinon.spy(request, 'write');
	//
	// 	this.request.returns(request);
	//
	// 	api.post(params, function() { });
	//
	// 	assert(write.withArgs(expected).calledOnce);
	// });

});


app.js
var express = require('express');
var app = express();

const oauth2 = simpleOauthModule.create({
	client: {
		id: process.env.DIGIKEY_CLIENT_ID,
		secret: process.env.DIGIKEY_CLIENT_SECRET,
		secretParamName: client_secret,
		idParamName: client_id
	},
	auth: {
		tokenHost: 'https://sso.digikey.com',
		tokenPath: '/as/token.oauth2'
		authorizePath: '/as/authorization.oauth2'
	},
});

// Authorization uri definition
const authorizationUri = oauth2.authorizationCode.authorizeURL({
	redirect_uri: process.env.DIGIKEY_REDIRECT_URI,
	scope: 'grant_type',
	state: 'authorization_code',
});

// Initial page redirecting to DigiKey
app.get('/auth', (req, res) => {
	console.log(authorizationUri);
	res.redirect(authorizationUri);
});

// Callback service parsing the authorization token and asking for the access token
app.get('/callback', (req, res) => {
	const code = req.query.code;
	const options = {
		code,
	};

	oauth2.authorizationCode.getToken(options, (error, result) => {
		if (error) {
			console.error('Access Token Error', error.message);
			return res.json('Authentication failed');
		}

		console.log('The resulting token: ', result);
		const token = oauth2.accessToken.create(result);

		return res
			.status(200)
			.json(token);
	});
});

app.get('/success', (req, res) => {
	res.send('');
});

app.get('/', (req, res) => {
	res.send('Hello<br><a href="/auth">Log in with DigiKey</a>');
});

app.listen(3000, () => {
	console.log('Express server started on port 3000'); // eslint-disable-line
});


keystone-route-example.js
var keystone = require('keystone');
var keystone = require('keystone');
var async = require('async');
var User = keystone.list('User').model;
var Token = keystone.list('Token').model;

// INDEX ROUTE SETUP
// DigiKey API
// app.all('/digikey/auth', keystone.middleware.api, routes.views.apis.digikey);
// app.all('/digikey/callback?', keystone.middleware.api, routes.views.apis.digikey);

// TOKEN MODEL EXAMPLE
// Token.add({
// 	token: { type: String, initial: true, required: true },
// 	refreshToken: { type: String, initial: true },
// 	tokenType: { type: String },
// 	expiresIn: { type: Types.Number },
// 	expiresAt: { type: Types.Datetime },
// 	service: {
// 		type: Types.Select,
// 		options: [
// 			{ value: 'octopart', label: 'Octopart' },
// 			{ value: 'digikey', label: 'Digikey' },
// 			{ value: 'twitter', label: 'Twitter' },
// 			{ value: 'linkedin', label: 'LinkedIn' },
// 			{ value: 'github', label: 'Github' }
// 		],
// 		emptyOption: true,
// 		index: true,
// 		initial: true,
// 		required: true
// 	},
// 	state: { type: Types.Select, options: 'draft, published, archived', default: 'published', index: true }
// });
//
// Token.relationship({ ref: 'User', path: 'users', refPath: 'tokens' });

const oauth2 = require('simple-oauth2').create({
	client: {
		id: process.env.DIGIKEY_CLIENT_ID,
		secret: process.env.DIGIKEY_CLIENT_SECRET,
		secretParamName: 'client_secret',
		idParamName: 'client_id'
	},
	auth: {
		tokenHost: 'https://sso.digikey.com',
		tokenPath: '/as/token.oauth2',
		authorizeHost: 'https://sso.digikey.com',
		authorizePath: '/as/authorization.oauth2'
	},
	options: {
		useBasicAuthorizationHeader: false,
		useBodyAuth: true
	}
});

// Authorization uri definition
const authorizationUri = oauth2.authorizationCode.authorizeURL({
	redirect_uri: process.env.DIGIKEY_REDIRECT_URI
});

exports = module.exports = function (req, res) {

	var view = new keystone.View(req, res);
	var locals = res.locals;

	// Init locals
	locals.section = 'digikey-callback';
	locals.filters = {
		code: (req.query.code) ? req.query.code : undefined
	};
	locals.accessToken = {};
	locals.tokenId = null;

	view.on('init', function (next) {
		if(locals.filters.code){
			next();
		} else {
			return res.redirect(authorizationUri);
		}
	});

	// Grab an authorization code
	view.on('init', function (next) {
		const code = locals.filters.code;
		if(!req.user){
			console.error('You must be logged in to perform this action.');
			return next(new Error('You must be logged in to perform this action.'));
		}
		if(!code){
			console.error('Unable to find a valid DigiKey authorization code');
			return next(new Error('Unable to find a valid DigiKey authorization code.'));
		}
		const options = {
			code
		};
		oauth2.authorizationCode.getToken(options, (error, result) => {
			if (error) {
				console.error('Access Token Error', error.message);
				return next(error);
			}

			// console.log('The resulting token: ', result);
			const token = oauth2.accessToken.create(result);
			locals.accessToken = token;

			// Generate a new token in the database
			var newToken = new Token({
				token: locals.accessToken.token.access_token,
				refreshToken: locals.accessToken.token.refresh_token,
				tokenType: locals.accessToken.token.token_type,
				expiresIn: locals.accessToken.token.expires_in,
				expiresAt: locals.accessToken.token.expires_at,
				service: 'digikey',
				state: 'published'
			});
			newToken.save();

			// Add the token to the current user and save
			var updatedUser = User.where({
				_id: req.user.id,
				state: 'active'
			})
			.update({ 'tokens': [ newToken ] })
			.setOptions({ multi: false })
			.exec(function(err){
				if(err){
					next(err);
				}
				// Update success
				next();
			});
		});
	});

	// Handle refresing the token
	// TODO: Implement this
	view.on('init', function (next) {
		// Grab the logged in user's token for digikey
		// User.findOne({
		// 	_id: req.user.id
		// })
		// .populate('tokens')
		// .exec(function(err, res){
		//
		// });

		// Sample of a JSON access token (you got it through previous steps)
		// const tokenObject = {
		// 	'access_token': '<access-token>',
		// 	'refresh_token': '<refresh-token>',
		// 	'expires_in': '7200'
		// };
		//
		// // Create the access token wrapper
		// const token = oauth2.accessToken.create(tokenObject);
		//
		// // Check if the token is expired. If expired it is refreshed.
		// if (token.expired()) {
		// 	// Callbacks
		// 	token.refresh((error, result) => {
		// 		token = result;
		// 	})
		//
		// 	// Promises
		// 	token.refresh()
		// 	.then((result) => {
		// 		token = result;
		// 	});
		// }
		next();
	});

	// Render the view
	view.render(function(err) {
		if(err) {
			return res.apiError('error', err);
		}
		res.apiResponse({
			token: locals.accessToken.token
		});
	});
};


index.js
var request = require('request');
var url = require('url');
var querystring = require('querystring');

// You can find more details about the Digikey API at https://api-portal.digikey.com/.

function flatten(arr) {
	var tmp = Array.prototype.concat.apply([], arr);
	return tmp.some(Array.isArray) ? flatten(tmp) : tmp;
}

function encodeFilters(filters) {
	var result = flatten(Object.keys(filters).map(function(k) {
		var v = Array.prototype.concat.apply([], filters[k]);
		return v.map(function(vi) {
			if (vi instanceof Object) {
				return Object.keys(vi).map(function(ki) {
					return k + '[' + ki + ']=' + querystring.escape(vi[ki]); // e.g. slice[field1]=1:5
				});
			} else {
				return k + '[]=' + querystring.escape(vi); // e.g. show[]=foo
			}
		});
	}));
	return result;
}

var DigiKeyNode = function(accessToken, apiVersion) {
	var self = this;

	self.clientId = process.env.DIGIKEY_CLIENT_ID;
	self.clientSecret = process.env.DIGIKEY_CLIENT_SECRET;
	self.redirectUri = process.env.DIGIKEY_CALLBACK_URL;
	self.apiVersion = (apiVersion) ? apiVersion : 'v1';
	self.accessToken = accessToken;
	self.refreshToken = null;
	self.expiresIn = 2592000;

	if(!self.accessToken){
		console.error('You must provide a valid access token. See https://api-portal.digikey.com/start for more.');
		return new Error('You must provide a valid access token. See https://api-portal.digikey.com/start for more.');
	}

	// Try an alternate env variable
	if(!self.redirectUri){
		if(process.env.DIGIKEY_REDIRECT_URI){
			self.redirectUri = process.env.DIGIKEY_REDIRECT_URI;
		}
	}

	self.country = 'US'; // See ISO Code http://www.nationsonline.org/oneworld/country_code_list.htm
	self.lang = 'en'; //"en", "de", "fr", "ko", "zhs", "zht", "it", "es", "he"
	self.locale = 'us'; // "ad", "ae", "af", "ag", "ag", "ai", "al", "am", "an", "ao", "aq", "ar", "as",
											// "at", "au", "aw", "ax", "az", "ba", "bb", "bd", "be", "bf", "bg", "bh", "bi",
											// "bj", "bm", "bn", "bo", "br", "bs", "bt", "bv", "bw", "by", "bz", "ca", "cc",
											// "cd", "cf", "cg", "ch", "ci", "ck", "cl", "cm", "cn", "co", "cr", "cs", "cu",
											// "cv", "cw", "cx", "cy", "cz", "de", "dj", "dk", "dm", "do", "dz", "ec", "ee",
											// "eg", "eh", "er", "es", "et", "fi", "fj", "fk", "fm", "fo", "fr", "ga", "gb",
											// "gd", "ge", "gf", "gh", "gi", "gl", "gm", "gn", "gp", "gq", "gr", "gs", "gt",
											// "gu", "gw", "gy", "hk", "hm", "hn", "hr", "ht", "hu", "id", "ie", "il", "in",
											// "io", "iq", "ir", "is", "it", "jm", "jo", "jp", "ke", "kg", "kh", "ki", "km",
											// "kn", "kp", "kr", "kw", "ky", "kz", "la", "lb", "lc", "li", "lk", "lr", "ls",
											// "lt", "lu", "lv", "ly", "ma", "mc", "md", "me", "mg", "mh", "mk", "ml", "mm",
											// "mn", "mo", "mp", "mq", "mr", "ms", "mt", "mu", "mv", "mw", "mx", "my", "mz",
											// "na", "nc", "ne", "nf", "ng", "ni", "nl", "no", "np", "nr", "nu", "nz", "om",
											// "pa", "pe", "pf", "pg", "ph", "pk", "pl", "pm", "pn", "pr", "ps", "pt", "pw",
											// "py", "qa", "re", "ro", "rs", "ru", "rw", "sa", "sb", "sc", "sd", "se", "sg",
											// "sh", "si", "sj", "sk", "sl", "sm", "sn", "so", "sr", "st", "sv", "sx", "sy",
											// "sz", "tc", "td", "tf", "tg", "th", "tj", "tk", "tl", "tm", "tn", "to", "tp",
											// "tr", "tt", "tv", "tw", "tz", "ua", "ug", "uy", "um", "us", "uz", "va", "vc",
											// "ve", "vg", "vi", "vn", "vu", "wf", "ws", "ye", "yt", "yu", "za", "zm", "zw"
	self.currency = 'usd'; //"usd", "cad", "jpy", "gbp", "eur", "hkd", "sgd", "twd", "krw", "aud", "nzd", "inr", "dkk", "nok", "sek", "ils"

	// Valid Endpoints
	// https://api.digikey.com/services/keywordsearch/v1/search
	// https://api.digikey.com/services/basicsearch/v1/search
	var send = function(service, path, params, filters, cb) {
		if (typeof filters === 'function') {
			cb = filters; // skip filters
		} else if (filters) {
			// Filters aren't used at the moment.
			// params = params.concat(encodeFilters(filters));
		}

		var opt = {
			method: 'POST',
			headers: {
				'x-digikey-locale-shiptocountry': self.country,
				'x-digikey-locale-currency': self.currency,
				'x-digikey-locale-site': self.locale,
				'x-digikey-locale-language': self.lang,
				authorization: self.accessToken,
				accept: 'application/json',
				'content-type': 'application/json',
				'x-ibm-client-id': self.clientId
			},
			url: url.format({
				protocol: 'https',
				host: 'api.digikey.com',
				pathname: '/services/' + service + '/' + self.apiVersion + '/' + path
			}),
			body: params,
			json: true
		};
		return request.get(opt,
			cb ? function(err, res, body) {
				if (err){
					cb(err);
				} else if (res.statusCode != 200) {
					cb(new Error(body.message));
				} else {
					cb(null, body);
				}
			} : null);
	};

	['basic', 'keyword'].forEach(function(name) {
		// The params are attached to the body of the request.
		// Example params
		// {
		// 	PartPreference: 'CT',  //Optional parameter for PartPreference which can be CT, DKR or TR. (Cut tape, Digi-Key Reel, Tape and Reel)
		//	Quantity: 25,
		//  PartNumber: 'P5555-ND'
		// }
		// filters = N/A at he moment
		self[name + 'Search'] = function(params, cb) {
			return send(name + 'search', 'search', params, {}, cb);
		};
	});

	// Handles refreshing the access token
	// See https://api-portal.digikey.com/app_overview#refreshToken
	self.refreshToken = function(refreshToken, expiresIn){
		// A token exists
		if(!refreshToken){
			var err = 'Unable to find a valid Digikey token.';
			console.error(err);
			return new Error(err);
		}

		self.expiresIn = (expiresIn) ? expiresIn : self.expiresIn;

		// JSON access token
		const tokenObject = {
			'access_token': self.accessToken,
			'refresh_token': self.refreshToken,
			'expires_in': self.expiresIn
		};

		// Create the access token wrapper
		const token = oauth2.accessToken.create(tokenObject);

		// Check if the token is expired. If expired it is refreshed.
		if (token.expired()) {
			// Promises
			return token.refresh();
		}
	};

	return self;
};

// https://sso.digikey.com/as/authorization.oauth2?response_type=code&client_id=123456789abcdefg&redirect_uri=https://my-new-app.example.com/code
exports.createV1 = function(accessToken) {
	return new DigiKeyNode(accessToken, 'v1');
};
